web/api/v1/api.go

```diff
@@ -492,14 +492,42 @@ func (api *API) labelNames(r *http.Request) apiFuncResult {
 	if err != nil {
 		return apiFuncResult{nil, &apiError{errorBadData, errors.Wrap(err, "invalid parameter 'end'")}, nil, nil}
 	}
+	var matcherSets [][]*labels.Matcher
+	for _, s := range r.Form["match[]"] {
+		matchers, err := parser.ParseMetricSelector(s)
+		if err != nil {
+			return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+		}
+		matcherSets = append(matcherSets, matchers)
+	}
 
 	q, err := api.Queryable.Querier(r.Context(), timestamp.FromTime(start), timestamp.FromTime(end))
 	if err != nil {
 		return apiFuncResult{nil, &apiError{errorExec, err}, nil, nil}
 	}
 	defer q.Close()
 
-	names, warnings, err := q.LabelNames()
+	var names []string
+	var warnings storage.Warnings
+	if len(r.Form["match[]"]) > 0 {
+		hints := &storage.SelectHints{
+			Start: timestamp.FromTime(start),
+			End:   timestamp.FromTime(end),
+			Func:  "series", // There is no series function, this token is used for lookups that don't need samples.
+		}
+
+		// Get all series which match matchers.
+		var sets []storage.SeriesSet
+		for _, mset := range matcherSets {
+			s := q.Select(false, hints, mset...)
+			sets = append(sets, s)
+		}
+		// TODO(yeya24): match these at TSDB level.
```

Comment:
Why do you want to do that? I'm not seeing how that would be notably any more efficient, without duplicating the entirety of the rather intricate postings matching logic which I don't think would be wise.

Reply:
Yeah, I agree. I will remove this comment.